std/zip.d

```diff
@@ -1137,59 +1131,117 @@ public:
 
-            auto zipfileCommentLength = getUshort(start + 20);
-            if (start + zipfileCommentLength + endOfCentralDirLength != data.length) continue;
-
-            enforce!ZipException(endrecOffset == to!uint(data.length),
-                                 "found more than one valid 'end of central dir record'");
+            if (start + EndOfCentralDirRecord.sizeof + fromLE!ushort(eocdr.commentLength) != data.length) continue;
 
+            enforce!ZipException(endrecOffset == data.length, "found more than one valid 'end of central dir record'");
             endrecOffset = start;
         }
 
-        enforce!ZipException(endrecOffset != to!uint(data.length),
+        enforce!ZipException(endrecOffset != data.length,
                              "found no valid 'end of central dir record'");
 
         return endrecOffset;
     }
 
+    private void checkForZip64(ulong offset)
+    {
+        // if we find a zip64locator signature at the correct place before the end of central dir
+        // record, we assume, that this archive is in zip64 format
+
+        if (offset < Zip64EndOfCentralDirLocatorRecord.sizeof) return;
+
+        ulong z64offset = offset - Zip64EndOfCentralDirLocatorRecord.sizeof;
+        _isZip64 = _data[z64offset .. z64offset + 4] == zip64EndOfCentralDirLocatorSignature;
+    }
+
+    private auto extractArchiveMember(ulong n, ulong next)
+    {
+        auto cfh = getRecord!CentralFileHeader(next);
+        enforce!ZipException(cfh.signature == centralFileHeaderSignature, "invalid directory signature");
+
+        ArchiveMember de = new ArchiveMember();
+        de._index = n;
+        de._madeVersion = fromLE!ushort(cfh.versionMade);
+        de._extractVersion = fromLE!ushort(cfh.versionExtract);
+        de.flags = fromLE!ushort(cfh.flags);
+        de._compressionMethod = cast(CompressionMethod) fromLE!ushort(cfh.compressionMethod);
+        de.time = cast(DosFileTime) fromLE!uint(cfh.time);
+        de._crc32 = fromLE!uint(cfh.crc32);
+        de._compressedSize = fromLE!uint(cfh.compressedSize);
+        de._expandedSize = fromLE!uint(cfh.expandedSize);
+        de.internalAttributes = fromLE!ushort(cfh.internalAttributes);
+        de._externalAttributes = fromLE!uint(cfh.externalAttributes);
+        de.offset = fromLE!uint(cfh.offset);
+
+        ulong namelen = fromLE!ushort(cfh.nameLength);
+        ulong extralen = fromLE!ushort(cfh.extraFieldLength);
+        ulong commentlen = fromLE!ushort(cfh.commentLength);
+
+        de.name = cast(string) getSegment(next + CentralFileHeader.sizeof, namelen);
+        de.extra = getSegment(next + CentralFileHeader.sizeof + namelen, extralen);
+        de.comment = cast(string) getSegment(next + CentralFileHeader.sizeof + namelen + extralen, commentlen);
+
+        extractAndVerifyArchiveMemberData(de);
+
+        _directory[de.name] = de;
+
+        return next + CentralFileHeader.sizeof + namelen + extralen + commentlen;
```

Comment:
Please do 
```
next += CentralFileHeader.sizeof;
de.name = ...;
next += namelen;
de.extra =  ...;
next += extralen;
...
return next;
```

Reply:
Also you could make `next` `ref` and make this `const` and `return de;` to be assigned to `_directory[de.name];` at the call site.